File: go/dsa/stack/impl/List/Stack.go
package stack

import (
	"errors"
)

// ListNode representa um nó da lista encadeada usada para a pilha.
type ListNode struct {
	data int       // Dado armazenado no nó
	next *ListNode // Ponteiro para o próximo nó na pilha
}

// Stack representa a estrutura de uma pilha, implementada com uma lista encadeada.
type Stack struct {
	top    *ListNode // Ponteiro para o nó no topo da pilha
	length int       // Quantidade de elementos na pilha
}

// NewStack cria e retorna uma nova pilha vazia.
func NewStack() *Stack {
	return &Stack{
		top:    nil,  // Inicialmente, o topo é nil, indicando que a pilha está vazia
		length: 0,    // A pilha começa com comprimento 0
	}
}

// Length retorna o número de elementos na pilha.
func (s *Stack) Length() int {
	return s.length // Retorna o valor da variável length que mantém o tamanho da pilha
}

// IsEmpty verifica se a pilha está vazia.
func (s *Stack) IsEmpty() bool {
	return s.length == 0 // Se length for 0, a pilha está vazia
}

// Push insere um novo elemento no topo da pilha.
func (s *Stack) Push(data int) {
	// Cria um novo nó com o dado fornecido
	newNode := &ListNode{data: data}

	// O próximo do novo nó passa a ser o nó atual do topo
	newNode.next = s.top

	// Atualiza o topo para o novo nó
	s.top = newNode

	// Incrementa o comprimento da pilha
	s.length++
}

// Pop remove e retorna o elemento do topo da pilha.
func (s *Stack) Pop() (int, error) {

	// Verifica se a pilha está vazia antes de tentar remover
	if s.IsEmpty() {
		return 0, errors.New("stack is empty") // Retorna um erro se a pilha estiver vazia
	}

	// Obtém o valor do topo
	result := s.top.data

	// Atualiza o topo para o próximo nó
	s.top = s.top.next

	// Decrementa o comprimento da pilha
	s.length--

	// Retorna o valor removido
	return result, nil
}

// Peek retorna o valor do topo da pilha sem removê-lo.
func (s *Stack) Peek() (int, error) {

	// Verifica se a pilha está vazia antes de tentar acessar o topo
	if s.IsEmpty() {
		return 0, errors.New("stack is empty") // Retorna um erro se a pilha estiver vazia
	}

	// Retorna o valor armazenado no topo da pilha
	return s.top.data, nil
}

File: csharp/src/dsa/stack/impl/list/Stack.cs
namespace dsa.stack.impl.list;

public class Stack
{
    // static void Main()
    // {
    //     Stack stack = new();

    //     stack.push(10);
    //     stack.push(15);
    //     stack.push(20);

    //     Console.WriteLine(stack.peek());
    //     stack.pop();

    //     Console.WriteLine(stack.peek());
    //     stack.pop();

    //     Console.WriteLine(stack.peek());
    //     stack.pop();

    //     // Tenta remover um elemento de uma pilha vazia, o que causará uma exceção
    //     try
    //     {
    //         stack.pop();
    //     }
    //     catch (Exception e)
    //     {
    //         Console.WriteLine(e.Message);
    //     }
    // }

    // A classe Stack tem duas propriedades: top e _length.
    // 'top' é um nó que aponta para o topo da pilha.
    // '_length' armazena o número de elementos na pilha.
    public ListNode top;
    public int _length;

    // A classe ListNode representa um nó na lista encadeada.
    public class ListNode
    {
        // Cada nó tem um valor de 'data' e um ponteiro 'next' para o próximo nó.
        public int data;
        public ListNode next;

        // O construtor da classe ListNode inicializa o valor de 'data' e define 'next' como null.
        public ListNode(int data)
        {
            this.data = data;
            this.next = null;
        }
    }

    // O construtor da classe Stack inicializa 'top' como null e '_length' como 0.
    public Stack()
    {
        top = null;
        _length = 0;
    }

    // O método 'length' retorna o número de elementos na pilha.
    public int length() => _length;

    // O método 'isEmpty' verifica se a pilha está vazia.
    public bool isEmpty() => _length == 0;

    // O método 'push' adiciona um novo elemento ao topo da pilha.
    public void push(int data)
    {
        // Cria um novo nó com o valor de 'data'.
        ListNode temp = new(data);
        // O novo nó aponta para o nó atual no topo da pilha.
        temp.next = top;
        // O novo nó se torna o topo da pilha.
        top = temp;
        // Incrementa o tamanho da pilha.
        _length++;
    }

    // O método 'pop' remove e retorna o elemento do topo da pilha.
    public int pop()
    {
        // Se a pilha estiver vazia, lança uma exceção.
        if (isEmpty())
            throw new Exception("Stack empty");

        // Armazena o valor do nó do topo.
        int result = top.data;
        // Move o topo para o próximo nó na pilha.
        top = top.next;
        // Decrementa o tamanho da pilha.
        _length--;
        // Retorna o valor do nó removido.
        return result;
    }

    // O método 'peek' retorna o valor do elemento do topo da pilha sem removê-lo.
    public int peek()
    {
        // Se a pilha estiver vazia, lança uma exceção.
        if (isEmpty())
            throw new Exception();

        // Retorna o valor do nó do topo.
        return top.data;
    }
}
